packages/SwingSet/src/kernel/virtualObjectManager.js

```diff
@@ -0,0 +1,342 @@
+   *    being fetched.
+   *
+   * @returns {*} an object representing the object's stored state.
+   */
+  function fetch(instanceKey) {
+    return JSON.parse(syscall.vatstoreGet(instanceKey));
+  }
+
+  /**
+   * Write an object's state to secondary storage.
+   *
+   * @param {string} instanceKey  The instance ID of the object whose state is
+   *    being stored.
+   * @param {*} rawData  A data object representing the state to be written.
+   */
+  function store(instanceKey, rawData) {
+    syscall.vatstoreSet(instanceKey, JSON.stringify(rawData));
+  }
+
+  /**
+   * Make a new kind of virtual object.
+   *
+   * @param {*} instanceMaker  A function of the form `instanceMaker(state)` that
+   *    will return a representative instance wrapped around the given state.
+   *
+   * @returns {*} a maker function that can be called to manufacture new
+   *    instance of this kind of object.  The parameters of the maker function
+   *    are those of the `initialize` method implemented in the representative
+   *    produced by the `instanceMaker` parameter.
+   */
+  function makeKind(instanceMaker) {
+    const kindID = `${allocateExportID()}`;
+    let nextInstanceID = 1;
+
+    function makeInnerSelf(instanceKey, rawData) {
+      const innerSelf = {
+        prev: undefined,
+        next: undefined,
+        instanceKey,
+        rawData,
+        dropData() {
+          store(instanceKey, innerSelf.rawData);
+          innerSelf.rawData = null;
+        },
+      };
+      return innerSelf;
+    }
+
+    function makeRepresentative(innerSelf, initArgs) {
+      function ensureState() {
+        if (!innerSelf.rawData) {
+          const instanceKey = innerSelf.instanceKey;
+          innerSelf = cache.lookup(instanceKey);
+          if (!innerSelf) {
+            innerSelf = makeInnerSelf(instanceKey, fetch(instanceKey));
+            cache.remember(innerSelf);
+          }
+        }
+        cache.refresh(innerSelf);
+      }
+
+      function wrapData() {
+        innerSelf.activeData = {};
+        for (const prop of Object.getOwnPropertyNames(innerSelf.rawData)) {
+          Object.defineProperty(innerSelf.activeData, prop, {
+            get: () => {
+              ensureState();
+              return m.unserialize(innerSelf.rawData[prop]);
+            },
+            set: value => {
+              ensureState();
+              innerSelf.rawData[prop] = m.serialize(value);
+            },
+          });
+        }
+        return harden(innerSelf.activeData);
+      }
+
+      if (!innerSelf.rawData && initArgs) {
+        innerSelf.rawData = {};
+        const tempInstance = instanceMaker(innerSelf.rawData);
+        tempInstance.initialize(...initArgs);
+        for (const prop of Object.getOwnPropertyNames(innerSelf.rawData)) {
+          innerSelf.rawData[prop] = m.serialize(innerSelf.rawData[prop]);
+        }
+      }
+      if (!innerSelf.activeData) {
+        wrapData();
```

Comment:
If we only create a new accessor-bearing `state` object (i.e. `innerSelf.activeData`) when the previous one had fallen off the cache, then the behavior code in one representative can compare its `state` object for identity against the `state` of a different representative to sense whether the cache has been busy or not, which is one of the subtle "no communication channels" requirements. I think we have to create a new `state` for every representative to avoid this.

Reply:
Yes, I think that would be cleaner.  It would entail moving the thing I called `activeData` out of the inner self and into the representative, which I think would work OK.  It sorts out an issue that is part of what got me into the way things are now, which is that the `marshal` library does not get along with objects whose properties are accessor functions (which caused things to blow up when I shifted from using `JSON.stringify/parse` as my placeholder serialization to using proper serialization via `marshal`.

Reply:
OK, that proved to be a trivial refactoring.